storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp

```diff
@@ -0,0 +1,313 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include "striped_btree_lockable_map.h"
+#include "btree_lockable_map.hpp"
+#include <algorithm>
+#include <cassert>
+#include <queue>
+
+namespace storage::bucketdb {
+
+namespace {
+
+constexpr uint8_t used_bits_of(uint64_t key) noexcept {
+    return static_cast<uint8_t>(key & 0b11'1111ULL);
+}
+
+}
+
+template <typename T>
+StripedBTreeLockableMap<T>::StripedBTreeLockableMap(uint8_t n_stripe_bits)
+    : _n_stripe_bits(n_stripe_bits),
+      _n_stripes(1ULL << _n_stripe_bits),
+      _stripes()
+{
+    assert(_n_stripe_bits > 0);
+    assert(_n_stripe_bits <= MaxStripeBits);
+    _stripes.reserve(_n_stripes);
+    for (size_t i = 0; i < _n_stripes; ++i) {
+        // TODO reduce initial sub-DB data store memory usage based on number of stripes
+        _stripes.emplace_back(std::make_unique<BTreeLockableMap<T>>());
+    }
+}
+
+template <typename T>
+StripedBTreeLockableMap<T>::~StripedBTreeLockableMap() = default;
+
+template <typename T>
+size_t StripedBTreeLockableMap<T>::stripe_of(key_type key) const noexcept {
+    assert(used_bits_of(key) >= _n_stripe_bits);
+    // Since bucket keys have count-bits at the LSB positions, we want to look at the MSBs instead.
+    return (key >> (64 - _n_stripe_bits));
```

Comment:
This might not work as intended if _n_stripe_bits is 0.

Reply:
`_n_stripe_bits` is asserted to be in the range [1, 8] at construction time, so this invariant should be maintained